src/Database/LSMTree/Internal/Index/Some.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Provides support for working with fence pointer indexes of unknown type.
+module Database.LSMTree.Internal.Index.Some
+(
+    IndexType (CompactIndex, OrdinaryIndex),
+    SomeIndex (SomeIndex),
+    headerLBS,
+    fromSBS,
+    SomeIndexAcc (SomeIndexAcc),
+    newWithDefaults
+)
+where
+
+import           Control.Arrow (second)
+import           Control.DeepSeq (NFData (..))
+import           Control.Monad.ST.Strict (ST)
+import           Data.ByteString.Lazy (LazyByteString)
+import           Data.ByteString.Short (ShortByteString)
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.Chunk (Chunk)
+import           Database.LSMTree.Internal.Entry (NumEntries)
+import           Database.LSMTree.Internal.Index (Index (..), IndexAcc (..))
+import qualified Database.LSMTree.Internal.Index.Compact as IndexCompact
+                     (fromSBS, headerLBS)
+import qualified Database.LSMTree.Internal.Index.CompactAcc as IndexCompact
+                     (new)
+import qualified Database.LSMTree.Internal.Index.Ordinary as IndexOrdinary
+                     (fromSBS, headerLBS)
+import qualified Database.LSMTree.Internal.Index.OrdinaryAcc as IndexOrdinary
+                     (new)
+import           Database.LSMTree.Internal.Page (NumPages, PageSpan)
+import           Database.LSMTree.Internal.Serialise (SerialisedKey)
+
+-- | The type of concrete index types specifically supported by this module.
+data IndexType
+    {-|
+        The type of compact indexes.
+
+        Compact indexes are designed to work with keys that are large (for
+        example, 32 bytes long) cryptographic hashes.
+
+        When using a compact index, it is vital that the
+        'Database.LSMTree.Internal.Serialise.Class.serialiseKey' function
+        satisfies the following law:
+
+        [Minimal size] @'Database.LSMTree.Internal.RawBytes.size'
+          ('Database.LSMTree.Internal.Serialise.Class.serialiseKey' x) >= 8@
+
+        Use 'Database.LSMTree.Internal.Serialise.Class.serialiseKeyMinimalSize'
+        to test this law.
+    -}
+    = CompactIndex
+      {-|
+          The type of ordinary indexes.
+
+          Ordinary indexes do not have any constraints on keys other than that
+          their serialised forms may not be 64 KiB or more in size.
+      -}
+    | OrdinaryIndex
+    deriving stock (Show, Eq)
+
+instance NFData IndexType where
+
+    rnf CompactIndex  = ()
+    rnf OrdinaryIndex = ()
+
+-- | The type of indexes.
+data SomeIndex = forall i . (Index i, NFData i) => SomeIndex i
```

Comment:
We could also go for explicitly listing the indexes:

```haskell
data SomeIndex = SomeCompactIndex CompactIndex | SomeOrdinaryIndex OrdinaryIndex
```

The family of available indexes is (indirectly) closed anyway, so maybe this is nicer to work with?